### ISSUE TITLE
Axolotl hanging on bench evals with fsdp

### ISSUE THREAD

**Issue by bsc001** [axolotl-ai-cloud/axolotl#2058]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

I m finetuning an LLAMA, and after an ep[och we need to launch the validation normally, but it is hanging

### Current behaviour

Axolotl is hanging for too long after the first epoch. I m finetuning a custom dataset on 2 RTX4090 using Qlora FSDP

`wandb: WARNING Saving files without folders. If you want to preserve subdirectories pass base_path to wandb.save, i.e. wandb.save("/mnt/folder/file.h5", base_path="/mnt")
[2024-11-14 13:13:49,394] [INFO] [axolotl.callbacks.on_train_begin:794] [PID:3439] [RANK:0] The Axolotl config has been saved to the WandB run under files.
  0%|                                                                                                                                         | 0/2792 [00:00<?, ?it/s]You're using a PreTrainedTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
{'loss': 0.9007, 'grad_norm': 18.620527267456055, 'learning_rate': 1.0000000000000002e-06, 'epoch': 0.0}                                                               
{'eval_loss': nan, 'eval_runtime': 68.9564, 'eval_samples_per_second': 8.527, 'eval_steps_per_second': 2.132, 'epoch': 0.0}                                            
  0%|                                                                                                                               | 1/2792 [01:16<5:39:51,  7.31s/it/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/transformers/generation/utils.py:2097: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cuda, whereas the model is on cpu. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cpu') before running `.generate()`.
  warnings.warn(
`

### Steps to reproduce

1. allocate 2 rtx 4090
2. launch the train using the config file
3. it will hang after first epoch

### Config yaml

```yaml
base_model: ./Llama-3.2-1B-Instruct
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer  
load_in_8bit: false
load_in_4bit: true
strict: false

seed: 42

datasets:
  - path: formated_partial_context.jsonl
    type: chat_template

    field_messages: conversations
    message_field_role: role
    message_field_content: content

dataset_prepared_path: last_run_prepared
val_set_size: 0.05
output_dir: ./outputs/out/qlora-Llama-3.2-1B-Instruct-seq4096-cuad-qa-partial

adapter: qlora
lora_model_dir:

sequence_len: 512
sample_packing: false
pad_to_sequence_len: true


evaluation_strategy : steps #epoch
eval_steps: 100
#evals_per_epoch: 1

eval_table_size: 3
eval_max_new_tokens: 512

do_causal_lm_eval: true
eval_causal_lm_metrics:
  - "context_matching"  # custom metric name
  - "perplexity" 


lora_r: 8
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project: llama3.2-1B-seq512-fine-tune-cuad-qa-partial
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model: checkpoint

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 4
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: false

early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
saves_per_epoch: 1
debug:
deepspeed:
weight_decay: 0.0
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: true
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
  activation_checkpointing: true
  pin_memory: false
special_tokens:
  pad_token: <|end_of_text|>
```

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.11

### axolotl branch-commit

official version 0.5.0

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by winglian** [comment]
@bsc001 is it hanging because it's a custom dataset or a custom evaluation? I see you have `context_matching` as a custom eval, and I'm not sure we can reproduce with that set. If you remove that eval, does it still hang? thanks!

**Comment by bsc001** [comment]
I run the training without custom metric and with default dataset and it halted

`wandb: WARNING Saving files without folders. If you want to preserve subdirectories pass base_path to wandb.save, i.e. wandb.save("/mnt/folder/file.h5", base_path="/mnt")
[2024-11-14 22:25:08,271] [INFO] [axolotl.callbacks.on_train_begin:794] [PID:3008] [RANK:0] The Axolotl config has been saved to the WandB run under files.
  0%|                                                                                                                            | 0/24700 [00:00<?, ?it/s]You're using a PreTrainedTokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.
{'loss': 1.4426, 'grad_norm': 1.3102000951766968, 'learning_rate': 1.0000000000000002e-06, 'epoch': 0.0}                                                   
{'eval_loss': nan, 'eval_runtime': 543.3375, 'eval_samples_per_second': 4.787, 'eval_steps_per_second': 2.394, 'epoch': 0.0}                               
  0%|                                                                                                                 | 1/24700 [09:09<43:34:08,  6.35s/it/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/transformers/generation/utils.py:2097: UserWarning: You are calling .generate() with the `input_ids` being on a device type different than your model's device. `input_ids` is on cuda, whereas the model is on cpu. You may experience unexpected behaviors or slower generation. Please make sure that you have put `input_ids` to the correct device by calling for example input_ids = input_ids.to('cpu') before running `.generate()`.
  warnings.warn(`

Here is the config :
`base_model: ./Llama-3.2-1B-Instruct
model_type: LlamaForCausalLM
tokenizer_type: AutoTokenizer  
load_in_8bit: false
load_in_4bit: true
strict: false

seed: 42

datasets:
  - path: tatsu-lab/alpaca
    type: alpaca

dataset_prepared_path: last_run_prepared
val_set_size: 0.05
output_dir: ./outputs/out/qlora-Llama-3.2-1B-Instruct-seq4096-cuad-qa-partial

adapter: qlora
lora_model_dir:

sequence_len: 512
sample_packing: false
pad_to_sequence_len: true


evaluation_strategy : steps #epoch
eval_steps: 100
#evals_per_epoch: 1

eval_table_size: 3
eval_max_new_tokens: 512

do_causal_lm_eval: true
eval_causal_lm_metrics:

lora_r: 8
lora_alpha: 16
lora_dropout: 0.05
lora_target_modules:
lora_target_linear: true
lora_fan_in_fan_out:

wandb_project: llama3.2-1B-seq512-fine-tune-cuad-qa-partial
wandb_entity:
wandb_watch:
wandb_name:
wandb_log_model: checkpoint

gradient_accumulation_steps: 4
micro_batch_size: 1
num_epochs: 4
optimizer: adamw_torch
lr_scheduler: cosine
learning_rate: 0.00001

train_on_inputs: false
group_by_length: false
bf16: auto
fp16:
tf32: false

early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true

warmup_steps: 10
saves_per_epoch: 1
debug:
deepspeed:
weight_decay: 0.0
fsdp:
  - full_shard
  - auto_wrap
fsdp_config:
  fsdp_limit_all_gathers: true
  fsdp_sync_module_states: true
  fsdp_offload_params: true
  fsdp_use_orig_params: false
  fsdp_cpu_ram_efficient_loading: true
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_transformer_layer_cls_to_wrap: LlamaDecoderLayer
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sharding_strategy: FULL_SHARD
  activation_checkpointing: true
  pin_memory: false
special_tokens:
  pad_token: <|end_of_text|>`

**Comment by NanoCode012** [comment]
Hey @bsc001 , thank you for the report. I can reproduce your issue. It seems to be due to fsdp error with bench evals. Could you swap to deepspeed? I found that it works for now while we try to find a fix.

**Comment by bsc001** [comment]
> Hey @bsc001 , thank you for the report. I can reproduce your issue. It seems to be due to fsdp error with bench evals. Could you swap to deepspeed? I found that it works for now while we try to find a fix.

Thank you, is deepspeed supporting qlora (i use qlora+fsdp for finetuning). Other question, do you think  the issue would happen if i switch to other GPUs ? what in FSDP is causing the halt ?
thabk you

**Comment by NanoCode012** [comment]
My findings show that the model is for some reason on cpu (with fsdp), while inputs are on gpu. Running generations on cpu is very slow. I've tried to move it to gpu / unwrap the model, but I couldn't get it working yet.

https://github.com/axolotl-ai-cloud/axolotl/blob/d9b71edf84887a4e63f862dec6f19d958a3c71dc/src/axolotl/utils/callbacks/__init__.py#L518-L525

>  is deepspeed supporting qlora

Yes, I did a run just now to be sure. Perplexity is logged. Moreover, the model trains fine and learn from dataset in my limited testing.

```yml
deepspeed: deepspeed_configs/zero1.json
```

**Comment by bsc001** [comment]
I believe it is due to : fsdp_cpu_ram_efficient_loading parameter, i will test without and get back to you

**Comment by NanoCode012** [comment]
> I believe it is due to : fsdp_cpu_ram_efficient_loading parameter, i will test without and get back to you

I tried that too and also `fsdp_offload_params`. Neither seems to solve it.

**Comment by NanoCode012** [comment]
Deepspeed z1-z2 would work. Z3 has a separate issue with peft model (if we properly unwrap the model).

**Comment by NanoCode012** [comment]
@bsc001 , I made a PR #2102 which should fix this issue. Are you able to test it? Currently, fsdp is very slow on eval, so I would recommend deepspeed in meantime.

**Comment by bsc001** [comment]
i will test and give you the feedback . thanks for the fix.
i want to use Qlora, so i ll need to use fsdp (siince deespeed is Z3 is not supporting peft).
